lib/client/alpn.go

```diff
@@ -0,0 +1,173 @@
+	AuthClient ALPNAuthClient
+
+	// Listener to be used to accept connections that will go trough the tunnel.
+	Listener net.Listener
+
+	// InsecureSkipTLSVerify turns off verification for x509 upstream ALPN proxy service certificate.
+	InsecureSkipVerify bool
+
+	// Protocol name
+	// Examples for databases: "postgres", "mysql"
+	// This protocol must map to a Teleport ALPN protocol [lib/srv/alpnproxy/common.alpnToALPNProtocol]
+	Protocol string
+
+	// PublicProxyAddr is public address of the proxy
+	PublicProxyAddr string
+
+	// ConnectionDiagnosticID contains the ID to be used to store Connection Diagnostic checks.
+	// Can be empty.
+	ConnectionDiagnosticID string
+
+	// RouteToDatabase contains the destination server that must receive the connection.
+	// Specific for database proxying.
+	RouteToDatabase proto.RouteToDatabase
+}
+
+// RunALPNAuthTunnel runs a local authenticated ALPN proxy to another service.
+// At least one Route (which defines the service) must be defined
+func RunALPNAuthTunnel(ctx context.Context, cfg ALPNAuthTunnelConfig) error {
+	alpnProtocol, err := alpn.ToALPNProtocol(cfg.Protocol)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	protocols := []alpn.Protocol{alpnProtocol}
+	if alpn.HasPingSupport(alpnProtocol) {
+		protocols = append(alpn.ProtocolsWithPing(alpnProtocol), protocols...)
+	}
+
+	var pool *x509.CertPool
+
+	alpnUpgradeRequired := alpnproxy.IsALPNConnUpgradeRequired(cfg.PublicProxyAddr, cfg.InsecureSkipVerify)
+
+	if alpnUpgradeRequired {
+		caCert, err := cfg.AuthClient.GetClusterCACert(ctx)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		pool = x509.NewCertPool()
+		if ok := pool.AppendCertsFromPEM(caCert.GetTLSCA()); !ok {
+			return trace.Errorf("failed to append cert from cluster's TLS CA Cert")
+		}
+	}
+
+	address, err := utils.ParseAddr(cfg.PublicProxyAddr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	tlsCert, err := getDatabaseUserCerts(ctx, cfg.AuthClient, cfg.RouteToDatabase, cfg.ConnectionDiagnosticID)
+	if err != nil {
+		return trace.BadParameter("failed to parse private key: %v", err)
+	}
+
+	lp, err := alpnproxy.NewLocalProxy(alpnproxy.LocalProxyConfig{
+		InsecureSkipVerify:      cfg.InsecureSkipVerify,
+		RemoteProxyAddr:         cfg.PublicProxyAddr,
+		Protocols:               protocols,
+		Listener:                cfg.Listener,
+		ParentContext:           ctx,
+		SNI:                     address.Host(),
+		Certs:                   []tls.Certificate{tlsCert},
+		RootCAs:                 pool,
+		ALPNConnUpgradeRequired: alpnUpgradeRequired,
+		Middleware:              nil,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	go func() {
+		defer cfg.Listener.Close()
+		if err := lp.Start(ctx); err != nil {
+			log.WithError(err).Info("ALPN proxy stopped.")
+		}
+	}()
+
+	return nil
+}
+
+func getDatabaseUserCerts(ctx context.Context, client ALPNAuthClient, routeToDatabase proto.RouteToDatabase, connectionDiagnosticID string) (tls.Certificate, error) {
```

Comment:
Nit: Why not return a pointer?
```suggestion
func getDatabaseUserCerts(ctx context.Context, client ALPNAuthClient, routeToDatabase proto.RouteToDatabase, connectionDiagnosticID string) (*tls.Certificate, error) {
```